File: benchmarks/poly/logparse/logparse.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"
)

func main() {
	t0 := time.Now()
	statuses := []string{"200", "200", "200", "404", "500"}
	cnt := make(map[string]int)
	lat := make(map[string]int)
	for i := 0; i < 150000; i++ {
		path := "/api/" + strconv.Itoa(i%50)
		status := statuses[i%5]
		line := "GET " + path + " " + status + " " + strconv.Itoa(i%250)
		f := strings.Split(line, " ")
		cnt[f[2]]++
		n, _ := strconv.Atoi(f[3])
		lat[f[1]] += n
	}
	keys := make([]string, 0, len(cnt))
	for k := range cnt {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	for _, k := range keys {
		fmt.Println(k + " " + strconv.Itoa(cnt[k]))
	}
	total := 0
	for _, v := range lat {
		total += v
	}
	fmt.Println(total)
	fmt.Fprintf(os.Stderr, "bench_ns=%d\n", time.Since(t0).Nanoseconds())
}

File: benchmarks/poly/logparse/logparse.pl
use strict; use warnings;
use Time::HiRes qw(clock_gettime CLOCK_MONOTONIC);
my $_t0 = clock_gettime(CLOCK_MONOTONIC);
my @statuses = ("200", "200", "200", "404", "500");
my (%cnt, %lat);
for my $i (0 .. 150000 - 1) {
    my $path = "/api/" . ($i % 50);
    my $status = $statuses[$i % 5];
    my $line = "GET " . $path . " " . $status . " " . ($i % 250);
    my @f = split / /, $line;
    $cnt{$f[2]}++;
    $lat{$f[1]} += int($f[3]);
}
for my $k (sort keys %cnt) {
    print $k . " " . $cnt{$k} . "\n";
}
my $total = 0;
$total += $_ for values %lat;
print $total, "\n";
printf STDERR "bench_ns=%.0f\n", (clock_gettime(CLOCK_MONOTONIC) - $_t0) * 1e9;
